cloudfront/wellcomelibrary.org/edge-lambda/src/api.ts

```diff
@@ -0,0 +1,63 @@
+import axios from 'axios';
+
+const apiBasePath = 'https://api.wellcomecollection.org/catalogue/v2';
+
+export type ApiQuery = {
+  query: string;
+  include: string[];
+};
+
+export type IdentifierType = {
+  id: string;
+  label: string;
+  type: 'IdentifierType';
+};
+
+export type Identifier = {
+  value: string;
+  identifierType: IdentifierType;
+  type: 'Identifier';
+};
+
+export type Work = {
+  type: 'Work' | 'Collection' | 'Section' | 'Series';
+  id: string;
+  title: string;
+  identifiers?: Identifier[];
+};
+
+export type CatalogueResultsList<Result = Work> = {
+  type: 'ResultList';
+  totalResults: number;
+  results: Result[];
+  pageSize: number;
+  prevPage?: string;
+  nextPage: string | undefined;
+};
+
+export type GetWorkResult = Work | Error;
+
+export async function* apiQuery(
+  query: ApiQuery
+): AsyncGenerator<Work, void, void> {
+  const url = `${apiBasePath}/works`;
+  const config = { params: query };
+
+  const apiResult = await axios.get(url, config);
+  const resultList = apiResult.data as CatalogueResultsList;
+
+  let nextPage = resultList.nextPage;
+  for await (const result of resultList.results) {
+    yield result;
+  }
+
+  while (nextPage) {
+    const apiResult = await axios.get(nextPage);
+    const resultList = apiResult.data as CatalogueResultsList;
+    nextPage = resultList.nextPage;
+
+    for await (const result of resultList.results) {
+      yield result;
+    }
+  }
+}
```

Comment:
Why does this get repeated? What about...
```typescript
let nextPage = true
while (nextPage) {
  const apiResult = await axios.get(nextPage);
  const resultList = apiResult.data as CatalogueResultsList;
  nextPage = resultList.nextPage;

  for await (const result of resultList.results) {
    yield result;
  }
}
```

Reply:
```ts
let nextPage = true
while (nextPage) {
  // nextPage is `true` in the first loop and doesn't work
  // I've tried a couple of ways and nothing I found is more 
  // satisfying than just putting up with the duplication
  const apiResult = await axios.get(nextPage);
  const resultList = apiResult.data as CatalogueResultsList;
  nextPage = resultList.nextPage;

  for await (const result of resultList.results) {
    yield result;
  }
}
```

Reply:
ohhhh I see, yes, fair enough